elyra/pipeline/catalog_connector.py

```diff
@@ -0,0 +1,489 @@
+#
+# Copyright 2018-2021 Elyra Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from abc import abstractmethod
+import hashlib
+from http import HTTPStatus
+import os
+from pathlib import Path
+from queue import Empty
+from queue import Queue
+from threading import Thread
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+
+from jupyter_core.paths import ENV_JUPYTER_PATH
+import requests
+from traitlets.config import LoggingConfigurable
+from traitlets.traitlets import CUnicode
+from traitlets.traitlets import Dict as DictTrait
+from traitlets.traitlets import Integer
+
+from elyra.metadata.metadata import Metadata
+
+
+class ComponentCatalogConnector(LoggingConfigurable):
+    """
+    Abstract class to model component_entry readers that can read components from different locations
+    """
+
+    # Data structure to encapsulate various capabilities & flags for a connector class
+    configuration = DictTrait(
```

Comment:
I believe @ptitzler requested a dictionary originally for extensibility purposes (some CCC's can add their own key-value pairs, for example). Is that right, Patrick?

Reply:
The value for `max_readers` is a system-level property and should not be something that other implementations can override (although there's really no way to prevent that).  If other CCC implementations need configurable traits, they are free to add them.  Those traits, though, will only be used by their implementation and not the general framework - as is the case here.

If we want to house this in a dictionary trait (which I've never seen in the ecosystem until now) that's fine, but I don't think these configurables should be privy to overrides, etc.

Reply:
Per discussion, we will revert the `max_readers` property to be a configurable Integer and remove this `configuration` Dictionary